File: libs/php/parser.php
<?php

/**
 * @file parser.php
 * @brief IPP project 1
 * @author Marek Gergel (xgerge01)
 * @date 2023-03-11
 * @details
 * Code parser class
 */

require_once 'params.php';
require_once 'error.php';

enum XMLTag: string
{
   case PROGRAM = "program";
   case INSTRUCTION = "instruction";
   case ARG1 = "arg1";
   case ARG2 = "arg2";
   case ARG3 = "arg3";
}

enum XMLAttribute: string
{
   case LANGUAGE = "language";
   case ORDER = "order";
   case OPCODE = "opcode";
   case TYPE = "type";
}

enum XMLAttributeValue: string
{
   case LABEL = "label";
   case VAR = "var";
   case TYPE = "type";
}

class Parser
{

   public $params;

   public function __construct(Params $params)
   {
      $this->params = $params;

      //open source file
      if ($this->params->source != STDIN) {
         $this->params->source = fopen($this->params->source, "r");
         if ($this->params->source == FALSE) error_exit(ErrorCodes::ERR_INPUT, "File not found or cannot be opened\n");
      }
   }

   public function __destruct()
   {
      //close source file
      if ($this->params->source != STDIN) fclose($this->params->source);
   }

   private function readLine()
   {
      while (($line = fgets($this->params->source)) !== FALSE) {
         //skip comments and empty lines
         if (str_contains($line, "#")) $line = substr($line, 0, strpos($line, "#"));
         if (!preg_match("/\w/", $line)) continue;

         return $line;
      }
      return FALSE;
   }

   private function parseHeader(DOMDocument $xml)
   {
      $line = $this->readLine();
      $line = trim($line);
      if (!preg_match("/^[.]ippcode23$/i", $line)) {
         error_exit(ErrorCodes::ERR_HEADER, "Invalid header\n");
      }

      $xml_program = $xml->createElement(XMLTag::PROGRAM->value);
      $xml_program->setAttribute(XMLAttribute::LANGUAGE->value, "IPPcode23");
      $xml->appendChild($xml_program);
   }

   private function parseProgram(DOMDocument $xml)
   {
      $xml_program = $xml->getElementsByTagName(XMLTag::PROGRAM->value)->item(0);

      $frame_charset = "(LF|TF|GF)";
      $name_charset = "([a-zA-Z_\-$(&amp;)%*!?][a-zA-Z0-9_\-$(&amp;)%*!?]*)";
      $type_charset = "(int|string|bool)";
      $type_charset_nil = "(int|string|bool|nil)";
      $const_charset =  "(nil@nil|int@([+-]?[0-9]+|nil)|string@([\\\\][0-9]{3}|[^\\\\])*|bool@(true|false|nil))";
      $regex_label = "/^$name_charset$/";
      $regex_var = "/^$frame_charset@$name_charset$/";
      $regex_symb = "/^(($frame_charset@$name_charset)|$const_charset)$/";
      $regex_type = "/^$type_charset$/";

      $loc = 0;

      //parse code
      while (($line = $this->readLine()) !== FALSE) {

         $loc++;
         $line = preg_replace("/&/", "&amp;", $line);
         $line = preg_replace("/</", "&lt;", $line);
         $line = preg_replace("/>/", "&gt;", $line);
         $line = trim($line);
         $line = preg_replace("/\s+/", " ", $line);
         $line = explode(" ", $line);

         //instruction
         $instruction = $line[0];
         if (!preg_match("/^(MOVE|CREATEFRAME|PUSHFRAME|POPFRAME|DEFVAR|CALL|RETURN|PUSHS|POPS|ADD|SUB|MUL|IDIV|LT|GT|EQ|AND|OR|NOT|INT2CHAR|STRI2INT|READ|WRITE|CONCAT|STRLEN|GETCHAR|SETCHAR|TYPE|LABEL|JUMP|JUMPIFEQ|JUMPIFNEQ|EXIT|DPRINT|BREAK)$/i", $instruction)) {
            error_exit(ErrorCodes::ERR_INSTRUCTION, "Invalid instruction\n");
         }

         //create instruction
         $xml_instruction = $xml->createElement(XMLTag::INSTRUCTION->value);
         $xml_instruction->setAttribute(XMLAttribute::ORDER->value, $loc);
         $xml_instruction->setAttribute(XMLAttribute::OPCODE->value, strtoupper($instruction));
         $xml_program->appendChild($xml_instruction);

         $xml_argv1 = NULL;
         $xml_argv2 = NULL;
         $xml_argv3 = NULL;

         //instruction params
         $instruction_params = array_slice($line, 1);
         $error_message = "Invalid number of instruction parameters\n";
         switch (count($instruction_params)) {
            case 0:
               if (!preg_match("/^(CREATEFRAME|PUSHFRAME|POPFRAME|RETURN|BREAK)$/i", $instruction)) {
                  error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, $error_message);
               }
               break;
            case 1:
               if (!preg_match("/^(DEFVAR|CALL|PUSHS|POPS|WRITE|LABEL|JUMP|EXIT|DPRINT)$/i", $instruction)) {
                  error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, $error_message);
               }
               break;
            case 2:
               if (!preg_match("/^(MOVE|NOT|INT2CHAR|READ|STRLEN|TYPE)$/i", $instruction)) {
                  error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, $error_message);
               }
               break;
            case 3:
               if (!preg_match("/^(ADD|SUB|MUL|IDIV|LT|GT|EQ|AND|OR|STRI2INT|CONCAT|GETCHAR|SETCHAR|JUMPIFEQ|JUMPIFNEQ)$/i", $instruction)) {
                  error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, $error_message);
               }
               break;
            default:
               error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, $error_message);
               break;
         }

         //label
         if (preg_match("/^(CALL|LABEL|JUMP|JUMPIFEQ|JUMPIFNEQ)$/i", $instruction)) {
            if (!preg_match($regex_label, $instruction_params[0])) error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, "Invalid label\n");
            $xml_argv1 = $xml->createElement(XMLTag::ARG1->value, $instruction_params[0]);
            $xml_argv1->setAttribute(XMLAttribute::TYPE->value, XMLAttributeValue::LABEL->value);
         }

         //var
         if (preg_match("/^(MOVE|DEFVAR|POPS|ADD|SUB|MUL|IDIV|LT|GT|EQ|AND|OR|NOT|INT2CHAR|STRI2INT|READ|CONCAT|STRLEN|GETCHAR|SETCHAR|TYPE)$/i", $instruction)) {
            if (!preg_match($regex_var, $instruction_params[0])) error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, "Invalid variable\n");
            $xml_argv1 = $xml->createElement(XMLTag::ARG1->value, $instruction_params[0]);
            $xml_argv1->setAttribute(XMLAttribute::TYPE->value, XMLAttributeValue::VAR->value);
         }

         //type
         if (preg_match("/^(READ)$/", $instruction)) {
            if (!preg_match($regex_type, $instruction_params[1])) error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, "Invalid type\n");
            $xml_argv2 = $xml->createElement(XMLTag::ARG2->value, $instruction_params[1]);
            $xml_argv2->setAttribute(XMLAttribute::TYPE->value, XMLAttributeValue::TYPE->value);
         }

         //symb(1)
         if (preg_match("/^(MOVE|PUSHS|ADD|SUB|MUL|IDIV|LT|GT|EQ|AND|OR|NOT|INT2CHAR|STRI2INT|WRITE|CONCAT|STRLEN|GETCHAR|SETCHAR|TYPE|JUMPIFEQ|JUMPIFNEQ|EXIT|DPRINT)$/i", $instruction)) {
            $index = preg_match("/^(PUSHS|WRITE|EXIT|DPRINT)$/i", $instruction) ? 0 : 1;
            if (!preg_match($regex_symb, $instruction_params[$index])) error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, "Invalid symbol\n");
            $xml_temp = $xml->createElement($index == 0 ? XMLTag::ARG1->value : XMLTag::ARG2->value, preg_replace("/^$type_charset_nil@/", "", $instruction_params[$index]));
            $xml_temp->setAttribute(XMLAttribute::TYPE->value, preg_match($regex_var, $instruction_params[$index]) ? XMLAttributeValue::VAR->value : preg_replace("/@.*$/", "", $instruction_params[$index]));
            $index == 0 ? $xml_argv1 = $xml_temp : $xml_argv2 = $xml_temp;
         }

         //symb2
         if (preg_match("/^(ADD|SUB|MUL|IDIV|LT|GT|EQ|AND|OR|STRI2INT|CONCAT|GETCHAR|SETCHAR|JUMPIFEQ|JUMPIFNEQ)$/i", $instruction)) {
            if (!preg_match($regex_symb, $instruction_params[2])) error_exit(ErrorCodes::ERR_INSTRUCTION_PARAMS, "Invalid symbol\n");
            $xml_argv3 = $xml->createElement(XMLTag::ARG3->value, preg_replace("/^$type_charset_nil@/", "", $instruction_params[2]));
            $xml_argv3->setAttribute(XMLAttribute::TYPE->value, preg_match($regex_var, $instruction_params[2]) ? XMLAttributeValue::VAR->value : preg_replace("/@.*$/", "", $instruction_params[2]));
         }

         if ($xml_argv1 !== NULL) $xml_instruction->appendChild($xml_argv1);
         if ($xml_argv2 !== NULL) $xml_instruction->appendChild($xml_argv2);
         if ($xml_argv3 !== NULL) $xml_instruction->appendChild($xml_argv3);
      }
   }

   public function getXML()
   {
      //create xml document
      $xml = new DOMDocument("1.0", "UTF-8");
      $xml->formatOutput = true;

      //parse header
      $this->parseHeader($xml);

      //parse program
      $this->parseProgram($xml);

      return $xml->saveXML();
   }
}
?>

File: libs/python/parser.py
# IPP 2023, project 2
# Author: Marek Gergel, xgerge01
# File: parser.py
# Description: Script xml structure parsing
# Date: 2023-04-07

from . import error,params,instruction,program
from collections import Counter
import xml.etree.ElementTree as et

# XML structure class with constants
class _XML:
    class Elem:
        program="program"
        instruction="instruction"
        arg1="arg1"
        arg2="arg2"
        arg3="arg3"

    class Attr:
        name="name"
        description="description"
        order="order"
        opcode="opcode"
        type="type"
        value="value"
        language="language"

    class AttrValue:
        IPPCODE23="IPPCODE23"
        label="label"
        type="type"
        var="var"
        int="int"
        bool="bool"
        string="string"
        float="float"
        nil="nil"

class parser:

    close_source = False

    def __init__(self):
        # Open source file 
        if not hasattr(params.source, "read"):
            try:
                params.source = open(params.source, "r")
                self.close_source = True
            except:
                error.exit(error.code.ERR_INPUT, f"Source file '{params.source}' does not exist or could not be read\n")

    def __del__(self):
        # Close source file 
        if self.close_source:
            params.source.close()

    def parseXML(self, prog: program.Program):
        # Parse xml structure
        try:
            tree = et.parse(params.source)
        except:
            error.exit(error.code.ERR_XML_FORMAT, "XML structure is not valid\n")

        # Get root element
        root = tree.getroot()

        # Check root element (program)
        if root.tag != _XML.Elem.program:
            error.exit(error.code.ERR_XML_SYNTAX, "Root element is not 'program'\n")
        if root.attrib[_XML.Attr.language].upper() != _XML.AttrValue.IPPCODE23:
            error.exit(error.code.ERR_XML_SYNTAX, "Code language is not 'IPPcode23'\n")

        # Check order attribute of child elements (instructions)
        try:
            values = [int(child.attrib[_XML.Attr.order]) for child in root]
        except:
            error.exit(error.code.ERR_XML_SYNTAX, "Order attribute does not exist\n")
        if any(not isinstance(value, int) for value in values):
            error.exit(error.code.ERR_XML_SYNTAX, "Order attribute value is not integer\n")
        if any(value < 1 for value in values):
            error.exit(error.code.ERR_XML_SYNTAX, "Order attribute value is less than 1\n")
        duplicates = [value for value, count in Counter(values).items() if count > 1]
        if duplicates:
            error.exit(error.code.ERR_XML_SYNTAX, "Order attribute value is not unique\n")

        # Sort child elements (instructions) by order attribute
        root[:] = sorted(root, key=lambda child: int(child.attrib[_XML.Attr.order]))

        # Check child elements (instructions)
        for child in root:
            instruction_in = self._parse_instruction(child)
            
            # Set label
            if isinstance(instruction_in, instruction.LABEL):
                # Label multiple definition
                if prog.label_is_defined(instruction_in.args[0][1]):
                    error.exit(error.code.ERR_XML_SEMANTIC, "Label already defined\n")
                prog.label_create(instruction_in.args[0][1], prog.instructions_count())
                
            # Add instruction to program
            prog.instruction_add(instruction_in)

    def _parse_instruction(self, root) -> instruction.Instruction:

        if root.tag != _XML.Elem.instruction:
            error.exit(error.code.ERR_XML_SYNTAX, f"Invalid instruction element '{root.tag}'\n")

        # Collect instruction arguments
        arg_count = len(root.findall("*"))
        args = {}

        if arg_count > 0:
            try:
                arg = root.find(_XML.Elem.arg1)
                args[0] = self._get_data_type(arg.attrib[_XML.Attr.type]), arg.text
            except:
                error.exit(error.code.ERR_XML_SYNTAX, f"Invalid argument element, '{_XML.Elem.arg1}' does not exist\n")

            if arg_count > 1:
                try:
                    arg = root.find(_XML.Elem.arg2)
                    args[1] = self._get_data_type(arg.attrib[_XML.Attr.type]), arg.text
                except:
                    error.exit(error.code.ERR_XML_SYNTAX, f"Invalid argument element, '{_XML.Elem.arg2}' does not exist\n")

                if arg_count > 2:
                    try:
                        arg = root.find(_XML.Elem.arg3)
                        args[2] = self._get_data_type(arg.attrib[_XML.Attr.type]), arg.text
                    except:
                        error.exit(error.code.ERR_XML_SYNTAX, f"Invalid argument element, '{_XML.Elem.arg3}' does not exist\n")

                    if arg_count > 3:
                        error.exit(error.code.ERR_XML_SYNTAX, f"Invalid instruction element, too many arguments\n")

        # Create instruction
        try:
            opcode = root.attrib[_XML.Attr.opcode].upper()
        except:
            error.exit(error.code.ERR_XML_SYNTAX, "Opcode attribute does not exist\n")

        match arg_count:
            case 0:
                if opcode not in ["CREATEFRAME", "PUSHFRAME", "POPFRAME", "RETURN", "BREAK", "CLEARS", "ADDS", "SUBS", "MULS", "IDIVS", "LTS", "GTS", "EQS", "ANDS", "ORS", "NOTS", "INT2CHARS", "STRI2INTS"]:
                    error.exit(error.code.ERR_XML_SYNTAX, f"Invalid instruction '{opcode}' with {arg_count} arguments\n")
            case 1:
                if opcode not in ["DEFVAR", "CALL", "PUSHS", "POPS", "WRITE", "LABEL", "JUMP", "EXIT", "DPRINT" , "JUMPIFEQS", "JUMPIFNEQS"]:
                    error.exit(error.code.ERR_XML_SYNTAX, f"Invalid instruction '{opcode}' with {arg_count} arguments\n")
            case 2:
                if opcode not in ["MOVE", "NOT", "INT2CHAR", "READ", "STRLEN", "TYPE"]:
                    error.exit(error.code.ERR_XML_SYNTAX, f"Invalid instruction '{opcode}' with {arg_count} arguments\n")
            case 3:
                if opcode not in ["ADD", "SUB", "MUL", "IDIV", "LT", "GT", "EQ", "AND", "OR", "STRI2INT", "CONCAT", "GETCHAR", "SETCHAR", "JUMPIFEQ", "JUMPIFNEQ"]:
                    error.exit(error.code.ERR_XML_SYNTAX, f"Invalid instruction '{opcode}' with {arg_count} arguments\n")

        match opcode:
            case "MOVE":
                return instruction.MOVE(opcode, args)
            case "CREATEFRAME":
                return instruction.CREATEFRAME(opcode, args)
            case "PUSHFRAME":
                return instruction.PUSHFRAME(opcode, args)
            case "POPFRAME":
                return instruction.POPFRAME(opcode, args)
            case "DEFVAR":
                return instruction.DEFVAR(opcode, args)
            case "CALL":
                return instruction.CALL(opcode, args)
            case "RETURN":
                return instruction.RETURN(opcode, args)
            case "PUSHS":
                return instruction.PUSHS(opcode, args)
            case "POPS":
                return instruction.POPS(opcode, args)
            case "ADD" | "ADDS":
                return instruction.ADD(opcode, args)
            case "SUB" | "SUBS":
                return instruction.SUB(opcode, args)
            case "MUL" | "MULS":
                return instruction.MUL(opcode, args)
            case "IDIV" | "IDIVS":
                return instruction.IDIV(opcode, args)
            case "LT" | "LTS":
                return instruction.LT(opcode, args)
            case "GT" | "GTS":
                return instruction.GT(opcode, args)
            case "EQ" | "EQS":
                return instruction.EQ(opcode, args)
            case "AND" | "ANDS":
                return instruction.AND(opcode, args)
            case "OR" | "ORS":
                return instruction.OR(opcode, args)
            case "NOT" | "NOTS":
                return instruction.NOT(opcode, args)
            case "INT2CHAR" | "INT2CHARS":
                return instruction.INT2CHAR(opcode, args)
            case "STRI2INT" | "STRI2INTS":
                return instruction.STRI2INT(opcode, args)
            case "READ":
                return instruction.READ(opcode, args)
            case "WRITE":
                return instruction.WRITE(opcode, args)
            case "CONCAT":
                return instruction.CONCAT(opcode, args)
            case "STRLEN":
                return instruction.STRLEN(opcode, args)
            case "GETCHAR":
                return instruction.GETCHAR(opcode, args)
            case "SETCHAR":
                return instruction.SETCHAR(opcode, args)
            case "TYPE":
                return instruction.TYPE(opcode, args)
            case "LABEL":
                return instruction.LABEL(opcode, args)
            case "JUMP":
                return instruction.JUMP(opcode, args)
            case "JUMPIFEQ" | "JUMPIFEQS":
                return instruction.JUMPIFEQ(opcode, args)
            case "JUMPIFNEQ" | "JUMPIFNEQS":
                return instruction.JUMPIFNEQ(opcode, args)
            case "EXIT":
                return instruction.EXIT(opcode, args)
            case "DPRINT":
                return instruction.DPRINT(opcode, args)
            case "BREAK":
                return instruction.BREAK(opcode, args)
            case "CLEARS":
                return instruction.CLEARS(opcode, args)
            case _: # Should never happen
                error.exit(error.code.ERR_XML_SYNTAX, f"Unknown opcode '{opcode}'\n")

    def _get_data_type(self, value_type: str):
        match value_type:
            case _XML.AttrValue.var:
                return program.Program.DataType.VAR
            case _XML.AttrValue.int:
                return int
            case _XML.AttrValue.bool:
                return bool
            case _XML.AttrValue.string:
                return str
            case _XML.AttrValue.float:
                return float
            case _XML.AttrValue.nil:
                return program.Program.DataType.NIL
            case _XML.AttrValue.label:
                return None
            case _XML.AttrValue.type:
                return None
            case _: # Should never happen
                error.exit(error.code.ERR_XML_SYNTAX, f"Unknown data type '{value_type}'\n")
